src/ChainSafe.Gaming.MetaMask.Unity/MetaMaskProvider.cs

```diff
@@ -0,0 +1,99 @@
+
+namespace ChainSafe.Gaming.MetaMask.Unity
+{
+    /// <summary>
+    /// Concrete implementation of <see cref="IMetaMaskProvider"/>.
+    /// </summary>
+    public class MetaMaskProvider : IMetaMaskProvider, ILifecycleParticipant
+    {
+        private readonly ILogWriter logWriter;
+
+        private readonly MetaMaskController metaMaskController;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetaMaskProvider"/> class.
+        /// </summary>
+        /// <param name="logWriter">Common Logger used for logging messages and errors.</param>
+        public MetaMaskProvider(ILogWriter logWriter)
+        {
+            this.logWriter = logWriter;
+
+            if (Application.isEditor || Application.platform != RuntimePlatform.WebGLPlayer)
+            {
+                this.logWriter.LogError("You need to build to WebGL platform to run Nethereum.Metamask.Unity");
+
+                return;
+            }
+
+            // Initialize Unity controller.
+            metaMaskController = Object.FindObjectOfType<MetaMaskController>();
+
+            if (metaMaskController == null)
+            {
+                GameObject controllerObj = new GameObject(nameof(MetaMaskController), typeof(MetaMaskController));
+
+                metaMaskController = controllerObj.GetComponent<MetaMaskController>();
+            }
+
+            Object.DontDestroyOnLoad(metaMaskController.gameObject);
+
+            metaMaskController.Initialize(this.logWriter);
+        }
+
+        /// <summary>
+        /// Implementation of <see cref="ILifecycleParticipant.WillStartAsync"/>.
+        /// Lifetime event method, called during initialization.
+        /// </summary>
+        /// <returns>async awaitable task.</returns>
+        public ValueTask WillStartAsync() => new ValueTask(Task.CompletedTask);
+
+        /// <summary>
+        /// Implementation of <see cref="IMetaMaskProvider.Connect"/>.
+        /// Called to connect to MetaMask.
+        /// </summary>
+        /// <returns>Connected account.</returns>
+        public async Task<string> Connect()
+        {
+            logWriter.Log("Connecting from Metamask...");
+
+            return await metaMaskController.Connect();
+        }
+
+        /// <summary>
+        /// Make JsonRPC requests using MetaMask.
+        /// </summary>
+        /// <param name="method">JsonRPC method name.</param>
+        /// <param name="parameters">JsonRPC request parameters.</param>
+        /// <typeparam name="T">Type of response result.</typeparam>
+        /// <returns>Response result.</returns>
+        public async Task<T> Request<T>(string method, params object[] parameters)
+        {
+            return await metaMaskController.Request<T>(method, parameters);
+        }
+
+        /// <summary>
+        /// Disconnect from MetaMask.
+        /// </summary>
+        /// <returns>Awaitable disconnect task.</returns>
+        public Task Disconnect()
```

Comment:
If there's no disconnection functionality available, then we can remove this method.

Reply:
The reason I kept it is because I would like Wallet Connect and Metamask providers inherit from one single class since they're both JsonRPC providers and have very similar functions

Reply:
we can provide one single point of abstraction for all wallet providers, since all they do is connect, disconnect and request Json RPC requests

Reply:
should be an issue on its own but this was kind of a prelude for it but I'll remove it if you sill think I should

Reply:
We already have Signer and TxExec abstractions that are essentially a breakdown of Wallet functionality. Some "Wallets" provide only Signer functionality, some provide both, some can also provide RPC Provider functionality. But I don't think we need more abstractions there. Let's keep it simple.
We can, however, try to transform those custom providers into a standardized IRpcClient, that's true. Not sure if it's possible though. This needs more research.
For now, I would remove this method just to keep the codebase clean.

Reply:
understood, done!